atecc608a_se.c

```diff
@@ -0,0 +1,212 @@
+    }
+
+    if (atcab_read_serial_number(buffer) != ATCA_SUCCESS)
+    {
+        atcab_release();
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_release() != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atecc608a_check_config_locked()
+{
+    bool config_locked;
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_is_locked(LOCK_ZONE_CONFIG, &config_locked) != ATCA_SUCCESS)
+    {
+        atcab_release();
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_release() != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    return config_locked? PSA_SUCCESS : PSA_ERROR_HARDWARE_FAILURE;
+}
+
+psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data, size_t data_size,
+                                         size_t *p_data_length)
+{
+    const size_t key_data_len = 65;
+    const uint16_t slot = key;
+    ATCA_STATUS ret = ATCA_SUCCESS;
+
+    if (data_size < key_data_len)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if ((ret = atcab_get_pubkey(slot, &p_data[1])) != ATCA_SUCCESS)
+    {
+        atcab_release();
+        return atecc608a_to_psa_error(ret);
+    }
+    p_data[0] = 4;
+    *p_data_length = key_data_len;
+
+    if (atcab_release() != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+#endif
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                       psa_algorithm_t alg,
+                                       const uint8_t *p_hash,
+                                       size_t hash_length,
+                                       uint8_t *p_signature,
+                                       size_t signature_size,
+                                       size_t *p_signature_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    const uint16_t key_id = key_slot;
+
+    /* We can only do ECDSA on SHA-256 */
+    /* PSA_ALG_ECDSA(PSA_ALG_SHA_256) */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
+    {
+        /* The driver only supports signing things of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    /* Signature will be returned here. Format is R and S integers in
+     * big-endian format. 64 bytes for P256 curve. */
```

Comment:
Since this function takes a `signature_size` argument and doesn't pass it down, check that it is large enough.